src/lib/descriptor/mod_test.rs

```diff
@@ -9,18 +8,60 @@ fn merge_env_both_empty() {
     let mut map1 = IndexMap::<String, EnvValue>::new();
     let mut map2 = IndexMap::<String, EnvValue>::new();
 
-    let output = merge_env(&mut map1, &mut map2);
+    let output = merge_env(&mut map1, &mut map2).expect("should have no cycle");
     assert_eq!(output.len(), 0);
 }
 
+#[test]
+fn merge_env_reorder() {
+    let mut map1 = IndexMap::new();
+    map1.insert("env2".to_owned(), EnvValue::Value("${env1}".to_owned()));
+
+    let mut map2 = IndexMap::new();
+    map2.insert("env1".to_owned(), EnvValue::Value("x".to_owned()));
+
+    let output = merge_env(&map1, &map2).expect("should have no cycle");
+    assert_eq!(output.len(), 2);
+    assert_eq!(output.get_index(0).unwrap().0.as_str(), "env1");
+    assert_eq!(output.get_index(1).unwrap().0.as_str(), "env2");
+}
+
+#[test]
+fn merge_env_reorder_internal() {
+    let mut map1 = IndexMap::new();
+    map1.insert("env3".to_owned(), EnvValue::Value("${env2}".to_owned()));
+    map1.insert("env2".to_owned(), EnvValue::Value("${env1}".to_owned()));
+
+    let mut map2 = IndexMap::new();
+    map2.insert("env1".to_owned(), EnvValue::Value("x".to_owned()));
+
+    let output = merge_env(&map1, &map2).expect("should have no cycle");
+    assert_eq!(output.len(), 3);
+    assert_eq!(output.get_index(0).unwrap().0.as_str(), "env1");
+    assert_eq!(output.get_index(1).unwrap().0.as_str(), "env2");
+    assert_eq!(output.get_index(2).unwrap().0.as_str(), "env3");
```

Comment:
i get why you want that, but if the env defined:
env3
env2
it should keep that order. it is up to the user while writing a makefile to put the env vars in order.

Reply:
Let us assist the user in being able to reorder the environment because this is what we already kind of do (on `merge_env`, we replace existing values). This is because we're handling configuration, not code. After all, I think in all cases, if I define a variable - independent of the order. I'd want them to be appropriately handled. This is how tools like `terraform` (https://www.terraform.io/language/values/outputs) work.

This also simplifies implementation, as we do a single passthrough and topologically sort.

This change is backward compatible, so old workflows won't break (evident by the CI not breaking) but instead enhances the current capabilities by allowing a more declarative approach.

Reply:
I think that, in general, if someone writes:

```
ENV2 = "${ENV3}"
ENV3 = "2"
```

They did not mean a global variable `ENV3`, then reassigned the value. I think this most likely leads to more confusion than it helps. The general mindset is that if I define an env variable in my block, then my program will use that variable. Otherwise, it might lead to unexpected results - and frustration - instead.

Reply:
I'd happily add a comment about that behavior in the `README.md`, but I think it's a valuable addition.

Reply:
maybe writing it in the README is a good idea. I have a long section with subsections on the env loading since its so complex and got so many capabilities already (env file, env on cli, os env, env block, profiles, env on task level, env scripts, .... )